src/ProvenanceTransformer.cpp

```diff
@@ -0,0 +1,452 @@
+#include "ProvenanceTransformer.h"
+#include "AstVisitor.h"
+
+namespace souffle {
+
+/**
+ * Helper functions
+ */
+const std::string& identifierToString(const AstRelationIdentifier& name) {
+    std::stringstream ss;
+    ss << name;
+    return *(new std::string(ss.str()));
+}
+
+inline AstRelationIdentifier makeRelationName(const AstRelationIdentifier& orig, const std::string& type, int num = -1) {
+    auto newName = new AstRelationIdentifier(identifierToString(orig));
+    newName->append(type);
+    if (num != -1) {
+        newName->append((const std::string&) std::to_string(num));
+    }
+
+    return *newName;
+}
+
+void addAttrAndArg(AstRelation* rel, AstAttribute* attr, AstAtom* head, AstArgument* arg) {
```

Comment:
It didn't seem more convenient since no variables are shared between the two halves. Unless I'm missing something, I think using the two original functions would be cleaner.